src/sites/experiments.cubing.net/cubing.js/svg-state-editor/index.ts

```diff
@@ -0,0 +1,201 @@
+            ...this.pieces.get(orbitName)[piece],
+            [orientation]: facelet,
+          };
+        }
+      }
+    }
+  }
+
+  getFaceletByOrientation(piece: Piece, orientation: number) {
+    return piece[orientation];
+  }
+
+  getPieceByFacelet({position, type}: Facelet) {
+    return this.pieces.get(type)[position];
+  }
+
+  swapFacelets(facelet1: Facelet, facelet2: Facelet) {
+    const temp = facelet1.element.style.fill;
+    facelet1.element.style.fill = facelet2.element.style.fill;
+    facelet2.element.style.fill = temp;
+  };
+
+  async twist(facelet: Facelet) {
+    const piece = this.getPieceByFacelet(facelet);
+
+    const { orbits } = (await this.puzzle.kpuzzle()).definition;
+
+    const { numOrientations } = orbits[facelet.type];
+
+    for (let i = 0; i < (numOrientations - 1); i++) {
+      const facelet = this.getFaceletByOrientation(piece, i);
+      const facelet2Orientation = (numOrientations + facelet.orientation + 1) % numOrientations;
+      this.swapFacelets(facelet, this.getFaceletByOrientation(piece, facelet2Orientation));
+    }
+  };
+
+  async swap(facelet1: Facelet, facelet2: Facelet) {
+    const piece1 = this.getPieceByFacelet(facelet1);
+    const piece2 = this.getPieceByFacelet(facelet2);
+
+    if (piece1 === piece2) {
+      return;
+    }
+
+    const offset = facelet2.orientation - facelet1.orientation;
+
+    const { orbits } = (await this.puzzle.kpuzzle()).definition;
+
+    const { numOrientations } = orbits[facelet1.type];
+
+    for (let i = 0; i < (numOrientations); i++) {
+      const facelet = this.getFaceletByOrientation(piece1, i);
+
+      const facelet2Orientation = (numOrientations + facelet.orientation + offset) % numOrientations;
+      this.swapFacelets(facelet, this.getFaceletByOrientation(piece2, facelet2Orientation));
+    }
+  };
+}
+
+class Facelet {
+  type: string;
+  position: number;
+  orientation: number;
+  element: HTMLOrSVGImageElement;
+
+  constructor(type: string, position: number, orientation: number) {
+    this.type = type;
+    this.orientation = orientation;
+    this.position = position;
+    this.element = document.getElementById(this.getId())! as HTMLOrSVGImageElement;
+    this.element.onclick = () => this.click();
+  }
+
+  getId() {
+    return `${this.type}-l${this.position}-o${this.orientation}`;
+  }
+
+  deselect() {
+    app.cube.selectedFacelet = null;
+    this.element.style.opacity = '1';
+  };
+
+  select() {
+    if (app.cube.selectedFacelet) {
+      app.cube.selectedFacelet.deselect();
+    }
+
+    app.cube.selectedFacelet = this;
+    this.element.style.opacity = '0.7';
+  };
+
+  click() {
+    switch (app.mode) {
+      case 'swap':
+        if (app.cube.selectedFacelet && app.cube.selectedFacelet.type == this.type) {
+          app.cube.swap(app.cube.selectedFacelet, this);
+
+          app.cube.selectedFacelet.deselect();
+        } else {
+          this.select();
+        }
+        break;
+      case 'twist':
```

Comment:
At some point I'd like to add right-click for clockwise twist as well, but that doesn't have to be in this PR.